src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+        }
+
+        public RegexToken InnerOpenParenToken { get; }
+        public RegexToken CaptureToken { get; }
+        public RegexToken InnerCloseParenToken { get; }
+
+        internal override int ChildCount => 7;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return OpenParenToken;
+            case 1: return QuestionToken;
+            case 2: return InnerOpenParenToken;
+            case 3: return CaptureToken;
+            case 4: return InnerCloseParenToken;
+            case 5: return Result;
+            case 6: return CloseParenToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```(?(group)result)```
+    /// </summary>
+    internal sealed class RegexConditionalExpressionGroupingNode : RegexConditionalGroupingNode
+    {
+        public RegexConditionalExpressionGroupingNode(
+            RegexToken openParenToken, RegexToken questionToken,
+            RegexGroupingNode grouping,
+            RegexExpressionNode result, RegexToken closeParenToken)
+            : base(RegexKind.ConditionalExpressionGrouping, openParenToken, questionToken, result, closeParenToken)
+        {
+            Debug.Assert(grouping != null);
+            Grouping = grouping;
+        }
+
+        internal override int ChildCount => 5;
+
+        public RegexGroupingNode Grouping { get; }
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return OpenParenToken;
+            case 1: return QuestionToken;
+            case 2: return Grouping;
+            case 3: return Result;
+            case 4: return CloseParenToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// Base type of all regex primitives that start with \
+    /// </summary>
+    internal abstract class RegexEscapeNode : RegexPrimaryExpressionNode
+    {
+        protected RegexEscapeNode(RegexKind kind, RegexToken backslashToken) : base(kind)
+        {
+            Debug.Assert(backslashToken.Kind == RegexKind.BackslashToken);
+            BackslashToken = backslashToken;
+        }
+
+        public RegexToken BackslashToken { get; }
+    }
+
+    /// <summary>
+    /// Base type of all regex escapes that start with \ and some informative character (like \v \t \c etc.).
+    /// </summary>
+    internal abstract class RegexTypeEscapeNode : RegexEscapeNode
+    {
+        protected RegexTypeEscapeNode(RegexKind kind, RegexToken backslashToken, RegexToken typeToken)
+            : base(kind, backslashToken)
+        {
+            TypeToken = typeToken;
```

Comment:
Could assert only one character in token?

Reply:
May have overlooked this suggestion too: Could assert only one character in token?

---
In reply to: [178450265](https://github.com/dotnet/roslyn/pull/23984#discussion_r178450265) [](ancestors = 178450265)

Reply:
That assert would unfortunatley not always be valid.  For example, in error conditions when teh tree ends with `\`.  Then you can an escape with a missing type token of length 0.

Reply:
Thanks for the clarification

---
In reply to: [210061116](https://github.com/dotnet/roslyn/pull/23984#discussion_r210061116) [](ancestors = 210061116)